### PR TITLE
750. Number Of Corner Rectangles

### PR DESCRIPTION
**Pull request by altayhunoglu**
Issue: https://github.com/spiralgo/algorithms/issues/271
The older solution did not work. I add the current fastest one with 10 ms.

### CODE REVIEW

**Comment by ErdemT09** [comment]
I will write an explanation here when I understand:

**Comment by ErdemT09** [comment]
I have actually came across this question while I was watching the mock Google interview by the guy from AlgoExpert with this competitive programmer.  https://youtu.be/EuPSibuIKIg (Watching the video itself is not useful though.)
This question might be an important coding question then.

File: src/main/java/algorithms/curated170/medium/NumberOfCornerRectangles.java [hunk 1]
```diff
@@ -13,21 +13,24 @@ public static void main(String[] args) {
...
-      
-      }
-      return count;
-      
- }
+   public int countCornerRectangles(int[][] grid) {
+        int totalCount = 0;
+        int rows = grid.length;
+        int cols = grid[0].length;
+        int[][] counts = new int[rows][cols];
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/NumberOfCornerRectangles.java:20]: Why don't we set the size to int[cols][cols]? Don't we only assign values to cols here?

**Comment by ErdemT09** [comment]
There is also this enumarating solution.
```
 public int _countCornerRectangles(int[][] grid) {
        int ans = 0;
        for (int i = 0; i < grid.length - 1; i++) {
            for (int j = i + 1; j < grid.length; j++) {
                int counter = 0;
                for (int k = 0; k < grid[0].length; k++) {
                    if (grid[i][k] == 1 && grid[j][k] == 1) counter++;
                }
                if (counter > 0) ans += counter * (counter - 1) / 2;
            }
        }
        return ans;
    }
```
Basically, we compare every single row with each other and add the rectangles that we find. This has a time complexity of `O(  row^2 * col )`, whereas the current solution has a complexity of `O(  col^2 * row )`.

File: src/main/java/algorithms/curated170/medium/NumberOfCornerRectangles.java (hunk 1 above)
Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/NumberOfCornerRectangles.java:20]: Yep, you are absolutely right. Somehow, it passes all LeetCode test cases. 
It means test cases are not rigorous enough.
Thanks for your attention. We fix it.

**Comment by altayhunoglu** [comment]
> There is also this enumarating solution.
> 
> ```
>  public int _countCornerRectangles(int[][] grid) {
>         int totalCount= 0;
>         for (int i = 0; i < grid.length - 1; i++) {
>             for (int j = i + 1; j < grid.length; j++) {
>                 int count = 0;
>                 for (int k = 0; k < grid[0].length; k++) {
>                     if (grid[i][k] == 1 && grid[j][k] == 1) count++;
>                 }
>                 if (count > 0) totalCount+= count * (count - 1) / 2;
>             }
>         }
>         return totalCount;
>     }
> ```
> 
> Basically, we compare every single row with each other and add the rectangles that we find. This has a time complexity of `O( row^2 * col )`, whereas the current solution has a complexity of `O( col^2 * row )`.

image

**Comment by ErdemT09** [comment]
Should we add the slower row-first based approach too?

**Comment by altayhunoglu** [comment]
> Should we add the slower row-first based approach too?

If you think it is necessary, of course we should add.

**Comment by altayhunoglu** [comment]
> > > Should we add the slower row-first based approach too?
> > 
> > 
> > If you think it is necessary, of course we should add.
> 
> I think we should, it has its own special sides as it is summing all numbers up to a certain count. Besides, it has a spatial complexity of O ( 1 ) and doesn't exchange that much speed for that.

Oke, it is done.

**Review by ErdemT09 — APPROVED** [review]

**Pull request merged**